core: mm: fix check_pa_matches_va() panic on unlocked paged memory

etienne-lms opened the pull request:
Change check_pa_matches_va() to not panic on memory references
on fully pageable (unlocked) memory.

This change makes tee_mmu_user_va2pa_attr() (and vm_va2pa()) to
return a null physical address with TEE_SUCCESS on unlocked
pageable memory since these may change of physical page.

An alternative is to inspect the pager to check is there is a
matching physical page but in this case one should lock pager
or at least the page, while va and pa are loaded and
check_pa_matches_va() returns. I think it's not worth the gain.

TODO: rename straight tee_mmu_user_va2pa_attr() to vm_va_info(pa, attr)
and update the inline comments stating vm_va2pa() is deprecated.

Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jforissier reviewed (commented):
- Pager code is not my area of expertise so pardon the obvious question perhaps... Does this fix an actual problem? You're saying `check_pa_matches_va()` may panic, but does it actually happen with the current code? If so an example would be good.
- The TODO needs to be removed from the commit description. Either add another patch or just leave it for later.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -2106,7 +2106,7 @@ static void check_pa_matches_va(void *va, paddr_t pa)
 			res = vm_va2pa(to_user_mode_ctx(thread_get_tsd()->ctx),
 				       va, &p);
-			if (res == TEE_SUCCESS && pa != p)
+			if (res == TEE_SUCCESS && pa != p && p != 0)
```
ok

etienne-lms commented:
> Pager code is not my area of expertise so pardon the obvious question perhaps... Does this fix an actual problem? You're saying check_pa_matches_va() may panic, but does it actually happen with the current code? If so an example would be good.

Yes, I faced the issue while adding tests or traces that called `virt_to_phys()` and panicked on paged pages when `CFG_TEE_CORE_DEBUG=y`. Maybe I could add this in the commit log.

> The TODO needs to be removed from the commit description. Either add another patch or just leave it for later.

Seems you're fine with such a change. I'll add a commit.

etienne-lms commented:
> > The TODO needs to be removed from the commit description. Either add another patch or just leave it for later.
> > 
> Seems you're fine with such a change. I'll add a commit.

Actually this was a bad idea. It looks better as is: vm.h exports vm_pa2va() and vm_va2pa() and that's fine.
Better than exporting `vm_pa2va()` and `tee_mmu_user_va2pa_attr()` :-)

etienne-lms commented:
patch updated

jforissier commented:
`Acked-by: Jerome Forissier <jerome@forissier.org>` but I wouldn't mind @jenswi-linaro's opinion too.

jenswi-linaro commented on core/mm/vm.c:
```diff
@@ -1159,11 +1159,16 @@ static TEE_Result tee_mmu_user_va2pa_attr(const struct user_mode_ctx *uctx,
-
-			*pa = p | ((vaddr_t)ua & (granule - 1));
+			if (mobj_is_paged(region->mobj)) {
+				*pa = 0;
```
I'd prefer if we return an error code here. Perhaps `TEE_ERROR_NOT_SUPPORTED` to clearly set it apart from `TEE_ERROR_ACCESS_DENIED`. We could update `mobj_with_fobj_get_pa()` to return `TEE_ERROR_NOT_SUPPORTED` instead of `TEE_ERROR_GENERIC()` when `f->fobj->ops->get_pa` is `NULL`.

etienne-lms replied:
ok, i'll look at your suggestion.

etienne-lms commented:
I updated the P-R as per Jens' suggestion. The change is smaller: is the 2 first commits in the series, but I added more commits mainly to add a bit of description on the modified function.
Note I did not test these changes yet. I'll give a try soon :)

jforissier commented on core/include/mm/vm.h:
```diff
@@ -87,18 +87,33 @@ TEE_Result vm_buf_to_mboj_offs(const struct user_mode_ctx *uctx,
+ * @va: Output virtual address reference
+ *
+ * Return TEE_ERROR_ACCESS_DENIED if the memory is not mapped in user context
+ * Return TEE_ERROR_NOT_SUPPORTED if mobj refers to pageable memory
```
Which `mobj`?

etienne-lms replied:
:) sorry!

jforissier commented on core/include/mm/vm.h:
```diff
@@ -87,18 +87,33 @@ TEE_Result vm_buf_to_mboj_offs(const struct user_mode_ctx *uctx,
- * Interface is deprecated, use virt_to_phys() instead.
- *---------------------------------------------------------------------------*/
+/*
+ * vm_va2pa - Translate a physical address to a virtual address in a user contex
```
You swapped physical and virtual.

etienne-lms replied:
thanks

jforissier commented on core/include/mm/vm.h:
```diff
@@ -87,18 +87,33 @@ TEE_Result vm_buf_to_mboj_offs(const struct user_mode_ctx *uctx,
- * Interface is deprecated, use phys_to_virt() instead.
- *---------------------------------------------------------------------------*/
+/*
+ * vm_pa2va - Translate a virtual address to a physical address in a user contex
```
You swapped virtual and physical.

etienne-lms replied:
let's be consistent

jforissier commented on core/include/mm/vm.h:
```diff
@@ -116,12 +116,25 @@ TEE_Result vm_va2pa(const struct user_mode_ctx *uctx, void *ua, paddr_t *pa);
+/*
+ * vm_set_ctx - Set memory mapping context
+ *
+ * @ctx: Session context, if NULL user mapping is removed and ASID set to 0
```
Mentioning ASID in a non-arch-specific file?

etienne-lms replied:
True. No need to mention ASID. User mapping is ok?

jforissier replied:
Yes

jforissier reviewed (commented):
In commit "core: arm: mm: fix init value of local variables in address conversion": s/initializier/initializer/ and change subject to "core: arm: mm: initialize local variables" (you're not really fixing anything, just following the coding standard).

etienne-lms commented:
actually I messed up the inline comments and there are thing still unclear.
I'll refine and propose a new series.

jforissier commented:
TBH I find many of the added comments useless, i.e., they do not provide any information that is not obvious already.

etienne-lms commented:
Right! and you already told me that :-)
True all these comments don't add anything but this specific return value deserves a description.
I'll update...
And thanks for your feedback, always welcome.

etienne-lms commented:
i've simplified the description: see fixup commit https://github.com/OP-TEE/optee_os/pull/4461/commits/b044e2cde2b0ea4e49acfaa512dea4393805e77b.
I'll squash this series and keep only the interesting changes.

jforissier commented:
@etienne-lms thanks, yes please squash and I'll review again.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -2106,6 +2106,8 @@ static void check_pa_matches_va(void *va, paddr_t pa)
 
 			res = vm_va2pa(to_user_mode_ctx(thread_get_tsd()->ctx),
 				       va, &p);
+			if (res == TEE_ERROR_NOT_SUPPORTED)
```
oups! i forgot to update this one!

etienne-lms commented:
I updated the P-R using TEE_ERROR_NO_DATA instead of TEE_ERROR_NOT_SUPPORTED.
Comments are welcome.

jforissier commented on core/arch/arm/include/mm/mobj.h:
```diff
@@ -47,6 +47,12 @@ static inline void *mobj_get_va(struct mobj *mobj, size_t offset)
 
+/*
+ * Get the physical address for a mobj memory cell
+ * Return TEE_ERROR_NO_DATA if mobj refers to volatile physical memory
```
"volatile" is not well defined here, I suggest: "refers to memory that is pageable and pager is enabled"

etienne-lms replied:
ok

jforissier reviewed (commented):
- For commit "core: arm: mm: initialize local variables":
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`
- For commit "core: arm: mobj: some mobjs may have no physical address":
With my comment addressed and the proper return value:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`
- For commit "core: mm: update description of vm.h exported functions":
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

jenswi-linaro commented:
> I updated the P-R using TEE_ERROR_NO_DATA instead of TEE_ERROR_NOT_SUPPORTED.
> Comments are welcome.

Why are you using `TEE_ERROR_NO_DATA` instead? I don't suppose it's super important, but `TEE_ERROR_NOT_SUPPORTED` seems like a more natural error code when a function isn't available.

jenswi-linaro commented on core/arch/arm/mm/mobj.c:
```diff
@@ -597,6 +597,9 @@ static TEE_Result mobj_with_fobj_get_pa(struct mobj *mobj, size_t offs,
+		return TEE_ERROR_NO_DATA;
+
 	if (!f->fobj->ops->get_pa)
 		return TEE_ERROR_GENERIC;
```
Why not just change this return code?

etienne-lms replied:
+1

jenswi-linaro commented on core/mm/vm.c:
```diff
@@ -1225,7 +1227,7 @@ TEE_Result vm_pa2va(const struct user_mode_ctx *uctx, paddr_t pa, void **va)
 	}
 
-	return TEE_ERROR_ACCESS_DENIED;
+	return TEE_ERROR_NO_DATA;
```
We do want to keep this as it means that it's an invalid virtual address.

etienne-lms replied:
In our case, the PA maybe be a valid address but vm_pa2va() can't resolve it because of the pager. So TEE_ERROR_ACCESS_DENIED seems not really accurate here.
That said, the sole caller of vm_pa2va() does not really care is the error is ACCESS_DENIED, NOT_SUPPORTED or NO_DATA.

jenswi-linaro replied:
OK, then I think it would be more clear to return the `va` on success or `NULL` on failure.

etienne-lms replied:
Ok, i'll update.

etienne-lms commented:
> Why are you using TEE_ERROR_NO_DATA instead? I don't suppose it's super important, but TEE_ERROR_NOT_SUPPORTED seems like a more natural error code when a function isn't available.

From this perspective NOT_SUPPORTED makes sense in a `*_get_pa()` helper function . 
Ok, i'll switch back to NO_SUPPORTED.

jenswi-linaro commented on core/include/mm/vm.h:
```diff
@@ -87,26 +83,20 @@ TEE_Result vm_buf_to_mboj_offs(const struct user_mode_ctx *uctx,
- * given the user context.
- * Interface is deprecated, use virt_to_phys() instead.
- *---------------------------------------------------------------------------*/
+/* This function is far less efficient than virt_to_phys() */
```
Instead of this comment it could be worth noting that this is a helper function for `virt_to_phys()` and shouldn't be used directly elsewhere.

jenswi-linaro commented on core/include/mm/vm.h:
```diff
@@ -87,26 +83,20 @@ TEE_Result vm_buf_to_mboj_offs(const struct user_mode_ctx *uctx,
- * Interface is deprecated, use phys_to_virt() instead.
- *---------------------------------------------------------------------------*/
-TEE_Result vm_pa2va(const struct user_mode_ctx *uctx, paddr_t pa, void **va);
+/* This function may be less efficient than phys_to_virt() */
```
Instead of this comment it could be worth noting that this is a helper function for phys_to_virt() and shouldn't be used directly elsewhere.

jenswi-linaro commented on core/include/mm/vm.h:
```diff
@@ -87,26 +83,20 @@ TEE_Result vm_buf_to_mboj_offs(const struct user_mode_ctx *uctx,
 
+/*
+ * Return TEE_SUCCESS or TEE_ERROR_ACCESS_DENIED when buffer exists unless
+ * what return another TEE_Result code.
```
Remove "what".

etienne-lms commented:
updated

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
@etienne-lms my R-b still apply.

etienne-lms commented:
I've split the series to 4 commits rather than 3. Ends to the same state.
I've applied the review tags on all commits. Please tell me if commit log and/or content is not consistent across commits.

jenswi-linaro commented:
Looks good.

jforissier commented:
Small glitch with Azure CI... build has completed in ~30 minutes but GitHub still thinks it's queued. Not a big deal.
In addition IBART seems to be down :-/

etienne-lms commented:
commit log for "core: mm: change vm_pa2va() to return a virtual address" is creepy:

> Change vm_pa2va() to return target virtual address or NULL **~~is~~ if** the
physical address cannot be resolve**d** which can happen when pager is
enabled and the target physical page belongs to pager page pool.
This change makes vm_pa2va() helper function simpler and **~~sole~~ its only** caller
doesn't **~~care~~**  differentiate error return codes.

May I change it before you merge?

jforissier commented:
@etienne-lms sure I'd like to get IBART to run before merging out of precaution @jbech-linaro

etienne-lms commented:
ok, i'll update.

etienne-lms commented:
Updated, tags applied, CI tests passed.
